File: src/friday/verification.py
from __future__ import annotations

import json
import platform
import re
from dataclasses import asdict
from pathlib import Path
from typing import Any, Callable

from agent_core import Agent, RunContext

from friday.agent_flow import GUARD_STOP_REASON, build_guarded_flow, inherit_guarded_run
from friday.config import (
    ModelConfig,
    build_model,
    load_model_config,
    output_token_limit,
)
from friday.model_options import DEFAULT_THINKING_EFFORT, thinking_request_kwargs
from friday.prompts import SECURITY_NOTES, VERIFIER_NOTES
from friday.storage import project_state_dir
from friday.tools import _shell_surface, allow_permissions_for_session, build_tools, pending_approval

VERIFIER_MAX_STEPS = 10000
def build_verifier(
    workspace: Path,
    config: ModelConfig | None = None,
    *,
    thinking_effort: str = DEFAULT_THINKING_EFFORT,
) -> tuple[Agent, RunContext]:
    root = workspace.resolve()
    friday_dir = project_state_dir(root)
    # One model, three roles: the verifier and the trace analyst reuse the
    # workspace's own model config, so a provider whose real ceiling is not
    # 1M never gets a window it cannot serve.
    config = config or load_model_config(root)
    system = platform.system()
    shell = "PowerShell" if system == "Windows" else "bash"
    tools = build_tools(root, friday_dir)
    agent = Agent(
        flow=build_guarded_flow(
            build_model(config),
            tools,
            chat_kwargs={
                "stream": False,
                # Thinking models burn output tokens on reasoning before writing
                # the verdict JSON. A tight cap like 900 can leave zero content
                # tokens, which fails parsing and blocks the whole turn.
                **output_token_limit(config, 4000),
                **thinking_request_kwargs(config.provider, thinking_effort),
                "tool_choice": "auto",
            },
        ),
        instructions=f"{SECURITY_NOTES}\n\n{VERIFIER_NOTES}\n\nWorkspace: {root}\nOS: {system}\nShell: {shell}",
    )
    context = agent.new_context()
    context.metadata["workspace"] = str(root)
    context.metadata["friday.model_config"] = asdict(config)
    context.metadata["friday.thinking_effort"] = thinking_effort
    # The verifier's job is to break the deliverable, which means running
    # builds, tests, and probes in manual mode without pausing the turn for
    # approval. Its commands are part of verification, not user work: exempt
    # it from permission prompts outright. Hard-denied commands (disk format,
    # credential exfiltration, encoded shell) stay blocked for every agent.
    allow_permissions_for_session(context)
    return agent, context


def verify_friday(goal: str, context: RunContext, start_event: int, *, force: bool = False) -> dict[str, Any] | None:
    events = [event.to_dict() for event in context.events[start_event:]]
    if not force and not needs_verification(events):
        return None
    workspace = Path(context.metadata["workspace"])
    session_id = str(context.metadata.get("session_id") or "")
    approval_result = _pending_approval_result(workspace, session_id)
    if approval_result:
        return approval_result
    context.emit("verification.start", category="verification", data={"goal_chars": len(goal)})
    config_data = context.metadata.get("friday.model_config")
    config = ModelConfig(**config_data) if isinstance(config_data, dict) else None
    thinking_effort = str(context.metadata.get("friday.thinking_effort") or DEFAULT_THINKING_EFFORT)
    verifier, verifier_context = build_verifier(workspace, config, thinking_effort=thinking_effort)
    verifier_context.usage = context.usage
    if context.on_observation is not None:
        def observe(event: Any) -> None:
            event.data["agent_role"] = "verifier"
            context.on_observation(event)

        verifier_context.on_observation = observe
    inherit_guarded_run(verifier_context, context)
    # The verifier is an extension of the main session: an approval it triggers
    # must land on this session's pending slot. Without the session id the
    # approval file is written to the shared slot, the post-run check below
    # misses it, and the suspended flow's empty answer is misreported as
    # broken verifier JSON. (Verification commands themselves no longer ask
    # for approval -- build_verifier grants them full access -- but keeping
    # the id makes any future approval path findable.)
    verifier_context.metadata["session_id"] = session_id
    try:
        history = [
            str(message.get("content") or "")[:1500]
            for message in context.get_messages()
            if message.get("role") == "user" and not message.get("friday_internal")
        ][-4:]
        if history and history[-1].strip() == goal.strip():
            history.pop()
        raw = verifier.chat(
            verification_prompt(goal, events, history),
            context=verifier_context,
            max_steps=VERIFIER_MAX_STEPS,
            stream=False,
        )
    except Exception as exc:
        return {"verdict": "inconclusive", "blocked": False, "error": True, "evidence": [], "feedback": f"Verifier failed: {exc}", "next_check": "", "passed": False, "required": True}
    approval_result = _pending_approval_result(workspace, session_id)
    if approval_result:
        return approval_result
    parsed = parse_verification(raw)
    guard_reason = verifier_context.metadata.get(GUARD_STOP_REASON)
    if isinstance(guard_reason, str):
        parsed["guard_stop_reason"] = guard_reason
    parsed["required"] = True
    return parsed


def record_verification(context: RunContext, result: dict[str, Any], on_verify: Callable[[dict[str, Any]], None] | None) -> None:
    context.emit("verification.result", category="verification", data=result)
    if on_verify:
        on_verify(result)


def needs_verification(events: list[dict[str, Any]]) -> bool:
    for event in events:
        data = event.get("data", {})
        if not isinstance(data, dict):
            continue
        name = str(data.get("name", ""))
        if event.get("type") == "tool.call" and name in {"Write", "Edit"}:
            return True
        if event.get("type") == "tool.call" and name == "Bash":
            args = data.get("arguments", {})
            command = str(args.get("command", "")) if isinstance(args, dict) else ""
            if bash_may_write(command):
                return True
    return False


def bash_may_write(command: str) -> bool:
    lowered = _shell_surface(command).lower()
    checked = re.sub(r"(?:(?:&|\*|\d+)\s*)?>{1,2}\s*(?:\$null\b|/dev/null\b|nul\b|&\d+\b)", "", lowered)
    return bool(re.search(r"\b(set-content|add-content|out-file|new-item|move-item|rename-item|rm|del|remove-item)\b|(^|[^><])>{1,2}(?![=>&])", checked))


def _pending_approval_result(workspace: Path, session_id: str = "") -> dict[str, Any] | None:
    approval = pending_approval(workspace, session_id=session_id)
    if not approval.get("pending"):
        return None
    return {
        "approval_required": True,
        "verdict": "inconclusive",
        "blocked": False,
        "evidence": [f"Pending approval for: {approval.get('command', '')}"],
        "feedback": "Approve or reject the pending command before continuing verification.",
        "passed": False,
        "required": True,
    }


def verification_prompt(goal: str, events: list[dict[str, Any]], user_history: list[str] | None = None) -> str:
    parts = [f"User goal:\n{goal}"]
    if user_history:
        parts.append(
            "Earlier user requirements (acceptance context, not proof):\n"
            + json.dumps(user_history, ensure_ascii=False, indent=2)
        )
    parts.extend(
        [
            "Independently verify the delivered workspace state by trying to break it. Use the delivery hints only to locate artifacts; they are not proof.",
            "Delivery hints:\n" + json.dumps(summarize_events(events), ensure_ascii=False, indent=2),
            'Return only JSON: {"verdict": "pass|repair|blocked|inconclusive", "evidence": ["criterion -> challenge -> outcome"], "feedback": "", "next_check": ""}',
        ]
    )
    return "\n\n".join(parts)


def summarize_events(events: list[dict[str, Any]]) -> list[dict[str, Any]]:
    summary = []
    for event in events:
        if event.get("type") != "tool.call":
            continue
        data = event.get("data", {})
        if not isinstance(data, dict):
            continue
        name = str(data.get("name") or "")
        if name not in {"Write", "Edit", "Bash"}:
            continue
        arguments = data.get("arguments", {})
        path = arguments.get("path") if isinstance(arguments, dict) else None
        hint = {"tool": name}
        if isinstance(path, str) and path:
            hint["path"] = path
        summary.append(hint)
    return summary[-20:]


def parse_verification(raw: str) -> dict[str, Any]:
    text = raw.strip()
    if not text:
        # An empty answer is not a malformed verdict: the verifier flow ended
        # without producing text (for example it suspended on a pending
        # approval it triggered). Name the real symptom instead of blaming the
        # JSON so the UI surfaces the actual blocker.
        return {
            "verdict": "inconclusive",
            "blocked": False,
            "error": True,
            "passed": False,
            "evidence": [],
            "feedback": "Verifier returned no output.",
            "next_check": "",
        }
    if text.startswith("```"):
        text = text.strip("`")
        text = text[text.find("{") :]
    start = text.find("{")
    end = text.rfind("}")
    if start != -1 and end != -1:
        try:
            value = json.loads(text[start : end + 1])
            if isinstance(value, dict):
                verdict = _verdict(value)
                feedback = str(value.get("feedback") or "").strip()
                next_check = str(value.get("next_check") or "").strip()
                if "verdict" not in value and verdict == "repair" and not next_check:
                    next_check = feedback
                return {
                    "verdict": verdict,
                    "blocked": verdict == "blocked",
                    "passed": verdict == "pass",
                    "evidence": value.get("evidence") if isinstance(value.get("evidence"), list) else [],
                    "feedback": feedback,
                    "next_check": next_check,
                }
        except json.JSONDecodeError:
            pass
    return {"verdict": "inconclusive", "blocked": False, "error": True, "passed": False, "evidence": [], "feedback": f"Verifier returned invalid JSON: {raw[:500]}", "next_check": ""}


def _verdict(value: dict[str, Any]) -> str:
    verdict = str(value.get("verdict") or "").strip().lower()
    if verdict in {"pass", "repair", "blocked", "inconclusive"}:
        return verdict
    if value.get("passed"):
        return "pass"
    if value.get("blocked"):
        return "blocked"
    return "repair" if value.get("feedback") else "inconclusive"

File: packages/harness/src/verification.ts
import { platform, release } from 'node:os'

import { Agent, RunContext, type AgentEvent, type ChatModel } from 'friday-agent-core'
import type { ExecutionBackend } from './plugin-api.js'

import type { ModelConfig } from './config.js'
import { modelFor } from './model.js'
import { promptTemplate } from './prompts.js'
import { buildVerifierTools } from './tools.js'

export type VerificationVerdict = 'pass' | 'repair' | 'blocked' | 'inconclusive'
export type VerificationResult = {
  verdict: VerificationVerdict
  passed: boolean
  blocked: boolean
  evidence: string[]
  feedback: string
  next_check: string
  required: true
  error?: boolean
  requests: number
  input_tokens: number | null
  output_tokens: number | null
  cached_tokens: number | null
  elapsed_ms: number
}

export async function verifyGoal(options: {
  /** Text-only goals may explicitly use the delivered answer as evidence. */
  answerEvidence?: string
  model?: ChatModel
  execution?: ExecutionBackend
  onEvent?: (event: AgentEvent) => void
  workspace: string
  config: ModelConfig
  thinking: string
  goal: string
  events?: readonly AgentEvent[]
  history?: readonly string[]
  signal?: AbortSignal
  toolSignal?: AbortSignal
}): Promise<VerificationResult> {
  const started = performance.now()
  const context = new RunContext()
  if (options.onEvent) context.onEvent = options.onEvent
  const shell = process.platform === 'win32' ? 'PowerShell' : 'sh'
  const instructions = [
    promptTemplate('SECURITY.md').trim(),
    promptTemplate('VERIFIER.md').trim(),
    `Workspace: ${options.workspace}\nOS: ${platform()} ${release()}\nShell: ${shell}`
  ].join('\n\n')
  const agent = new Agent({
    model: options.model ?? modelFor(options.config, options.thinking, 4_000),
    tools: buildVerifierTools(options.workspace, options.execution),
    instructions,
    maxSteps: 40,
    beforeStep: () => {
      if (!options.toolSignal?.aborted) return
      if (!context.messages.some(message => message.verifier_finishing)) {
        context.addMessage({
          role: 'system',
          content: 'The verification run has entered its finishing reserve. Do not call more tools. Return the strict verification JSON from the evidence already collected.',
          agent_internal: true,
          verifier_finishing: true
        })
      }
      return { tools: false as const }
    }
  }, context)
  try {
    const result = await agent.run(verificationPrompt(options.goal, options.events ?? [], options.history ?? []), {
      ...(options.signal ? { signal: options.signal } : {}),
      ...(options.toolSignal ? { toolSignal: options.toolSignal } : {})
    })
    let parsed = parseVerification(result.text)
    const successful = new Set(context.events.filter(event => event.type === 'tool.result' && event.data.is_error === false).map(event => String(event.data.tool_call_id)))
    if (parsed.verdict === 'pass' && (result.status !== 'done' || !parsed.evidence.length || !parsed.evidence.every(line => {
      const references = [...line.matchAll(/\[tool:([^\]]+)\]/g)].map(match => match[1]!)
      return references.length ? references.every(id => successful.has(id)) : !!options.answerEvidence && line.includes('[answer]')
    }))) {
      parsed = { ...parsed, verdict: 'inconclusive', passed: false, feedback: 'Pass rejected: each criterion needs a reference to a successful verifier tool result, or explicitly enabled answer evidence.', next_check: '' }
    }
    return {
      ...parsed,
      required: true,
      requests: context.usage.requests,
      input_tokens: context.usage.inputTokens,
      output_tokens: context.usage.outputTokens,
      cached_tokens: context.usage.cachedTokens,
      elapsed_ms: Math.round(performance.now() - started)
    }
  } catch (error) {
    if (options.signal?.aborted) throw error
    return {
      verdict: 'inconclusive', passed: false, blocked: false, evidence: [],
      feedback: `Verifier failed: ${error instanceof Error ? error.message : String(error)}`,
      next_check: '', required: true, error: true,
      requests: context.usage.requests,
      input_tokens: context.usage.inputTokens,
      output_tokens: context.usage.outputTokens,
      cached_tokens: context.usage.cachedTokens,
      elapsed_ms: Math.round(performance.now() - started)
    }
  }
}

export function parseVerification(raw: string): Omit<VerificationResult, 'required' | 'requests' | 'input_tokens' | 'output_tokens' | 'cached_tokens' | 'elapsed_ms'> {
  const match = raw.slice(raw.indexOf('{'), raw.lastIndexOf('}') + 1)
  try {
    const value: unknown = JSON.parse(match)
    if (!value || typeof value !== 'object' || Array.isArray(value)) throw new Error('verdict is not an object')
    const record = value as Record<string, unknown>
    const verdict = String(record.verdict || '').trim().toLowerCase()
    if (!['pass', 'repair', 'blocked', 'inconclusive'].includes(verdict)) throw new Error('unknown verdict')
    if (verdict === 'pass' && (!Array.isArray(record.evidence) || !record.evidence.some(item => typeof item === 'string' && item.trim()))) throw new Error('pass requires evidence')
    return {
      verdict: verdict as VerificationVerdict,
      passed: verdict === 'pass',
      blocked: verdict === 'blocked',
      evidence: Array.isArray(record.evidence)
        ? record.evidence.slice(0, 20).map(item => String(item).trim().slice(0, 1_000)).filter(Boolean)
        : [],
      feedback: typeof record.feedback === 'string' ? record.feedback.trim().slice(0, 4_000) : '',
      next_check: typeof record.next_check === 'string' ? record.next_check.trim().slice(0, 2_000) : ''
    }
  } catch {
    return {
      verdict: 'inconclusive', passed: false, blocked: false, evidence: [],
      feedback: raw.trim() ? `Verifier returned invalid JSON: ${raw.trim().slice(0, 500)}` : 'Verifier returned no output.',
      next_check: '', error: true
    }
  }
}

function verificationPrompt(goal: string, events: readonly AgentEvent[], history: readonly string[]): string {
  const parts = [`User goal:\n${goal.trim()}`]
  const earlier = history.map(value => value.trim()).filter(value => value && value !== goal.trim()).slice(-4)
  if (earlier.length) parts.push(`Earlier user requirements (acceptance context, not proof):\n${JSON.stringify(earlier, null, 2)}`)
  parts.push(
    'Independently verify the delivered workspace state by trying to break it. Use the delivery hints only to locate artifacts; they are not proof.',
    `Delivery hints:\n${JSON.stringify(deliveryHints(events), null, 2)}`,
    'Each pass evidence line must cite a tool result from THIS verification using [tool:tool_call_id]. A claim with no successful tool result is insufficient. Return only JSON: {"verdict":"pass|repair|blocked|inconclusive","evidence":["criterion -> challenge -> outcome [tool:call_id]"],"feedback":"","next_check":""}'
  )
  return parts.join('\n\n')
}

function deliveryHints(events: readonly AgentEvent[]): Array<Record<string, string>> {
  return events.flatMap(event => {
    if (event.type !== 'tool.call') return []
    const name = String(event.data.name || '')
    if (!['Write', 'Edit', 'Bash'].includes(name)) return []
    const args = event.data.arguments
    const path = args && typeof args === 'object' && !Array.isArray(args)
      ? (args as Record<string, unknown>).path
      : undefined
    return [{ tool: name, ...(typeof path === 'string' && path ? { path } : {}) }]
  }).slice(-20)
}
